spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -206,11 +219,14 @@ public StructType readSchema() {
     Broadcast<Table> tableBroadcast = sparkContext.broadcast(SerializableTable.copyOf(table));
 
     List<InputPartition<ColumnarBatch>> readTasks = Lists.newArrayList();
-    for (CombinedScanTask task : tasks()) {
-      readTasks.add(new ReadTask<>(
-          task, tableBroadcast, expectedSchemaString, caseSensitive,
-          localityPreferred, new BatchReaderFactory(batchSize)));
-    }
+    Tasks.foreach(tasks())
+        .stopOnFailure()
+        .executeWith(readTasksInitExecutorService)
+        .run(task -> {
+          readTasks.add(new ReadTask<>(
```

Comment:
This can't alter `readTasks` from multiple threads because array lists aren't thread-safe.

Reply:
Oh, that was a mistake. Fixed it!